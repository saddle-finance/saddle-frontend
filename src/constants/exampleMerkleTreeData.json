{
  "merkleRoot": "0x21aba301f38af8f5dfc161be38796812443a1749f6e0744de51e5c03781767fa",
  "allowedAccounts": {
      "0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC": {
          "proof": [
              "0xe9707d0e6171f728f7473c24cc0432a9b07eaaf1efed6a137a4a8c12c79552d9",
              "0xa216d5da9e6fad5b8d8a82e2df1aa2465b66fb408f2da71ff5fe6f272da5b11b"
          ],
          "flags": {
              "yearn": true,
              "curve": true,
              "yam": true
          }
      },
      "0x70997970C51812dc3A010C7d01b50e0d17dc79C8": {
          "proof": [
              "0x7b52b82e709191be4bae3bb8544d2a6b51d7440a1d15ba1f09e33bcf049e47a5",
              "0x4ed9d015110a35000ce5c94f94ccdc63653ddd26af11314d386ae5e65ef28c79"
          ],
          "flags": {
              "yearn": true,
              "curve": true,
              "yam": true
          }
      },
      "0xA51c1fc2f0D1a1b8494Ed1FE312d7C3a78Ed91C0": {
          "proof": [
              "0x00314e565e0574cb412563df634608d76f5c59d9f817e85966100ec1d48005c0",
              "0x4ed9d015110a35000ce5c94f94ccdc63653ddd26af11314d386ae5e65ef28c79"
          ],
          "flags": {
              "yearn": true,
              "curve": true,
              "yam": true
          }
      },
      "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266": {
          "proof": [
              "0x8a3552d60a98e0ade765adddad0a2e420ca9b1eef5f326ba7ab860bb4ea72c94",
              "0xa216d5da9e6fad5b8d8a82e2df1aa2465b66fb408f2da71ff5fe6f272da5b11b"
          ],
          "flags": {
              "yearn": true,
              "curve": true,
              "yam": false
          }
      }
  }
}