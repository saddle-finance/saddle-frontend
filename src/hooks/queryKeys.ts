export enum QueryKeys {
  RegistryAddress = "registryAddress",
  SidechainGaugeWeightData = "sidechainGaugeWeightData",
}
