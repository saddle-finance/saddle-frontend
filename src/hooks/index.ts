import { useAccount, useChainId, useProvider, useSigner } from "wagmi"

import { ChainId } from "../constants/networks"
import { useMemo } from "react"

export function useActiveWeb3React() {
  const chainId = useChainId() as ChainId
  const { address: account } = useAccount()
  const provider = useProvider()
  const { data: signer } = useSigner()

  const signerOrProvider = useMemo(() => signer || provider, [signer, provider])

  const activeWeb3React = useMemo(
    () => ({ chainId, account, signerOrProvider }),
    [chainId, account, signerOrProvider],
  )

  return activeWeb3React
}
