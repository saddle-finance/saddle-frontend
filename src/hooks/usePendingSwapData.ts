import {
  BLOCK_TIME,
  ChainId,
  IS_VIRTUAL_SWAP_ACTIVE,
  SWAP_TYPES,
} from "../constants"
import {
  BasicToken,
  BasicTokens,
  TokensContext,
} from "../providers/TokensProvider"
import { useCallback, useContext, useEffect, useState } from "react"

import { BigNumber } from "@ethersproject/bignumber"
import { Bridge } from "../../types/ethers-contracts/Bridge"
import { Event } from "ethers"
import { Zero } from "@ethersproject/constants"
import { omit } from "lodash"
import { useActiveWeb3React } from "./"
import { useBridgeContract } from "./useContract"

export interface PendingSwap {
  swapType: SWAP_TYPES
  settleableAtTimestamp: number
  secondsRemaining: number
  synthTokenFrom: BasicToken
  synthBalance: BigNumber
  tokenTo: BasicToken
  itemId: string
  transactionHash: string
  timestamp: number
  events: Array<SettlementEvent | WithdrawEvent>
}

interface SwapEvent {
  timestamp: number
  itemId: string
  transactionHash: string // used for deduping
  type: "settlement" | "withdraw"
}
export interface SettlementEvent extends SwapEvent {
  fromToken: BasicToken
  fromAmount: BigNumber
  toToken: BasicToken
  toAmount: BigNumber
  type: "settlement"
}
export interface WithdrawEvent extends SwapEvent {
  synthToken: BasicToken
  amount: BigNumber
  type: "withdraw"
}

type BridgeEventSettle = {
  requester: string
  itemId: BigNumber
  settleFrom: string
  settleFromAmount: BigNumber
  settleTo: string
  settleToAmount: BigNumber
  isFinal: boolean
}

type BridgeEventWithdraw = {
  requester: string
  itemId: BigNumber
  synth: string
  synthAmount: BigNumber
  isFinal: boolean
}

type BridgeEventGenericSwap = {
  requester: string
  itemId: BigNumber
  // other fields exist but we don't use them
}

type State = {
  pendingSwaps: Omit<PendingSwap, "events">[]
  settlements: { [itemId: string]: SettlementEvent[] }
  withdraws: { [itemId: string]: WithdrawEvent[] }
}

const VIRTUAL_SWAP_TOPICS = [
  "TokenToSynth",
  "SynthToToken",
  "TokenToToken",
] as const
const SETTLE = "Settle"
const WITHDRAW = "Withdraw"

const usePendingSwapData = (): PendingSwap[] => {
  const { account, chainId, library } = useActiveWeb3React()
  const bridgeContract = useBridgeContract()
  const tokens = useContext(TokensContext)
  const [state, setState] = useState<State>({
    pendingSwaps: [],
    settlements: {},
    withdraws: {},
  })
  const queryStartBlock =
    chainId === ChainId.HARDHAT
      ? 0
      : -1 * Math.round((365 * 24 * 60 * 60 * 1000) / BLOCK_TIME) // approx number of blocks in 1 year

  // update the secondsRemaining every 15s
  useEffect(() => {
    const timer = setInterval(() => {
      void library?.getBlock("latest").then(({ timestamp }) => {
        // block timestamp in secs
        setState((prevState) => {
          const shouldUpdateState = prevState.pendingSwaps.some(
            ({ secondsRemaining }) => secondsRemaining > 0,
          )
          if (!shouldUpdateState) return prevState
          return {
            ...prevState,
            pendingSwaps: prevState.pendingSwaps.map((swap) => {
              let secondsRemaining = Math.floor(
                swap.settleableAtTimestamp / 1000 - timestamp,
              )
              secondsRemaining = Math.max(secondsRemaining, 0)
              return { ...swap, secondsRemaining }
            }),
          }
        })
      })
    }, 15 * 1000)
    return () => {
      clearInterval(timer)
    }
  }, [library])
  const pendingSwapEventListener = useCallback(
    (...listenerArgs) => {
      const event = listenerArgs[listenerArgs.length - 1] as Event
      const itemIdArg = event.args?.itemId as BigNumber | null
      if (itemIdArg == null || bridgeContract == null || chainId == null) return
      const itemId = itemIdArg.toString()
      void fetchPendingSwapInfo(bridgeContract, event, tokens).then(
        (fetchedPendingSwap) => {
          if (fetchedPendingSwap == null) return
          setState((prevState) => {
            return {
              ...prevState,
              pendingSwaps: [
                ...prevState.pendingSwaps.filter(
                  ({ itemId: existingItemId }) => itemId !== existingItemId,
                ),
                fetchedPendingSwap,
              ],
            }
          })
        },
      )
    },
    [bridgeContract, chainId, setState, tokens],
  )
  const settleEventListener = useCallback(
    (...listenerArgs) => {
      const event = listenerArgs[listenerArgs.length - 1] as Event
      if (chainId == null) return
      void parseSettlementFromEvent(event, tokens).then((settlement) => {
        if (settlement == null) return
        setState((prevState) => {
          const pendingSwap = prevState.pendingSwaps.find(
            ({ itemId }) => itemId === settlement.itemId,
          )
          const newPendingSwapBalance = pendingSwap
            ? pendingSwap.synthBalance.sub(settlement.fromAmount)
            : Zero
          const prevSettlements = prevState.settlements[settlement.itemId] || []
          const txnsSet = new Set(
            prevSettlements.map(({ transactionHash }) => transactionHash),
          )
          if (newPendingSwapBalance.lte(Zero)) {
            // if this action empties the pending swap balance, remove it from state
            return removeItemFromState(prevState, settlement.itemId)
          } else if (!txnsSet.has(settlement.transactionHash)) {
            return {
              ...prevState,
              settlements: {
                ...prevState.settlements,
                [settlement.itemId]: prevSettlements.concat(settlement),
              },
            }
          } else {
            return prevState
          }
        })
      })
    },
    [chainId, setState, tokens],
  )
  const withdrawEventListener = useCallback(
    (...listenerArgs) => {
      const event = listenerArgs[listenerArgs.length - 1] as Event
      if (chainId == null) return
      void parseWithdrawFromEvent(event, tokens).then((withdraw) => {
        if (withdraw == null) return
        setState((prevState) => {
          const pendingSwap = prevState.pendingSwaps.find(
            ({ itemId }) => itemId === withdraw.itemId,
          )
          const newPendingSwapBalance = pendingSwap
            ? pendingSwap.synthBalance.sub(withdraw.amount)
            : Zero
          const prevWithdraws = prevState.withdraws[withdraw.itemId] || []
          const txnsSet = new Set(
            prevWithdraws.map(({ transactionHash }) => transactionHash),
          )
          if (newPendingSwapBalance.lte(Zero)) {
            // if this action empties the pending swap balance, remove it from state
            return removeItemFromState(prevState, withdraw.itemId)
          } else if (!txnsSet.has(withdraw.transactionHash)) {
            return {
              ...prevState,
              withdraws: {
                ...prevState.withdraws,
                [withdraw.itemId]: prevWithdraws.concat(withdraw),
              },
            }
          } else {
            return prevState
          }
        })
      })
    },
    [chainId, setState, tokens],
  )
  // attach listeners for events
  useEffect(() => {
    function attachPendingSwapEventListeners() {
      if (bridgeContract == null || account == null) return
      const virtualSwapTopicFilters = VIRTUAL_SWAP_TOPICS.map((topic) =>
        bridgeContract.filters[topic](account, null, null, null, null, null),
      )
      const settleFilter = bridgeContract.filters[SETTLE](
        account,
        null,
        null,
        null,
        null,
        null,
        null,
      )
      const withdrawFilter = bridgeContract.filters[WITHDRAW](
        account,
        null,
        null,
        null,
        null,
      )

      virtualSwapTopicFilters.forEach((filter) => {
        void bridgeContract.on(filter, pendingSwapEventListener)
      })
      bridgeContract.on(settleFilter, settleEventListener)
      bridgeContract.on(withdrawFilter, withdrawEventListener)
      return () => {
        virtualSwapTopicFilters.forEach((filter) => {
          void bridgeContract.off(filter, pendingSwapEventListener)
        })
        bridgeContract.off(settleFilter, settleEventListener)
        bridgeContract.off(withdrawFilter, withdrawEventListener)
      }
    }
    void attachPendingSwapEventListeners()
  }, [
    bridgeContract,
    account,
    settleEventListener,
    withdrawEventListener,
    pendingSwapEventListener,
  ])
  // initial fetch of events
  useEffect(() => {
    async function fetchExistingEvents() {
      if (!IS_VIRTUAL_SWAP_ACTIVE || !account || !bridgeContract || !chainId)
        return

      await Promise.all([
        fetchAndPopulatePendingSwaps(
          bridgeContract,
          setState,
          account,
          tokens,
          queryStartBlock,
        ),
        fetchAndPopulateWithdraws(
          bridgeContract,
          setState,
          account,
          tokens,
          queryStartBlock,
        ),
        fetchAndPopulateSettlements(
          bridgeContract,
          setState,
          account,
          tokens,
          queryStartBlock,
        ),
      ])
    }
    void fetchExistingEvents()
  }, [bridgeContract, setState, account, chainId, queryStartBlock, tokens])

  return state.pendingSwaps.map((swap) => {
    // merge swaps + settlements state
    return {
      ...swap,
      events: [
        ...(state.settlements[swap.itemId] || []),
        ...(state.withdraws[swap.itemId] || []),
      ].sort((a, b) => a.timestamp - b.timestamp),
    }
  })
}

async function fetchAndPopulatePendingSwaps(
  bridgeContract: Bridge,
  setState: (callback: (state: State) => State) => void,
  account: string,
  tokens: BasicTokens,
  startBlock: number,
): Promise<void> {
  // Step 1: create event filters
  const eventFilters = VIRTUAL_SWAP_TOPICS.map((topic) => {
    return bridgeContract.filters[topic](account, null, null, null, null, null)
  })
  // Step 2: run queries and get matching events
  let events
  events = await Promise.all(
    eventFilters.map((filter) =>
      bridgeContract.queryFilter(filter, startBlock, "latest"),
    ),
  )
  events = events.flat()

  // Step 3: fetch aync data via bridge contract
  let fetchedPendingSwaps = await Promise.all(
    events.map((event) => fetchPendingSwapInfo(bridgeContract, event, tokens)),
  )
  fetchedPendingSwaps = fetchedPendingSwaps.filter(Boolean)
  if (fetchedPendingSwaps.length === 0) return

  // Step 4: write to state
  setState((prevState) => {
    return {
      ...prevState,
      pendingSwaps: fetchedPendingSwaps as PendingSwap[],
    }
  })
}
async function fetchAndPopulateWithdraws(
  bridgeContract: Bridge,
  setState: (callback: (state: State) => State) => void,
  account: string,
  tokens: BasicTokens,
  startBlock: number,
): Promise<void> {
  // Step 1: create event filter
  const eventFilter = bridgeContract.filters[WITHDRAW](
    account,
    null,
    null,
    null,
    null,
  )

  // Step 2: run queries and get matching events
  const events = await bridgeContract.queryFilter(
    eventFilter,
    startBlock,
    "latest",
  )
  // Step 3: get each event's block
  const withdraws = await Promise.all(
    events.map((event) => parseWithdrawFromEvent(event, tokens)),
  )

  // Step 4: write to state
  if (withdraws.length === 0) return
  setState((prevState) => {
    const newState = { ...prevState }
    withdraws.forEach((withdraw) => {
      if (withdraw == null) return
      const existingWithdraws = newState.withdraws[withdraw.itemId]
      newState.withdraws[withdraw.itemId] = (existingWithdraws || []).concat(
        withdraw,
      )
    })
    return newState
  })
}
async function fetchAndPopulateSettlements(
  bridgeContract: Bridge,
  setState: (callback: (state: State) => State) => void,
  account: string,
  tokens: BasicTokens,
  startBlock: number,
): Promise<void> {
  // Step 1: create event filter
  const eventFilter = bridgeContract.filters[SETTLE](
    account,
    null,
    null,
    null,
    null,
    null,
    null,
  )

  // Step 2: run queries and get matching events
  const events = await bridgeContract.queryFilter(
    eventFilter,
    startBlock,
    "latest",
  )
  // Step 3: get each event's block
  const settlements = await Promise.all(
    events.map((event) => parseSettlementFromEvent(event, tokens)),
  )

  // Step 4: write to state
  if (settlements.length === 0) return
  setState((prevState) => {
    const newState = { ...prevState }
    settlements.forEach((settlement) => {
      if (settlement == null) return
      const existingSettlements = newState.settlements[settlement.itemId]
      newState.settlements[settlement.itemId] = (
        existingSettlements || []
      ).concat(settlement)
    })
    return newState
  })
}

async function parseSettlementFromEvent(
  event: Event,
  tokens: BasicTokens,
): Promise<SettlementEvent | null> {
  // Settlements are final
  const settlement = event.args as unknown as BridgeEventSettle // TODO would love to set this from the typechain type
  if (settlement == null || !tokens) return null
  const fromToken = tokens[settlement.settleFrom]
  const toToken = tokens[settlement.settleTo]
  if (fromToken == null || toToken == null) return null
  return event
    .getBlock()
    .then(({ timestamp }) => {
      return {
        fromAmount: settlement.settleFromAmount,
        fromToken,
        itemId: settlement.itemId.toString(),
        timestamp,
        toAmount: settlement.settleToAmount,
        toToken,
        transactionHash: event.transactionHash,
        type: "settlement" as const,
      }
    })
    .catch(() => null)
}

async function parseWithdrawFromEvent(
  event: Event,
  tokens: BasicTokens,
): Promise<WithdrawEvent | null> {
  const withdraw = event.args as unknown as BridgeEventWithdraw
  if (withdraw == null || !tokens) return null
  const synthToken = tokens[withdraw.synth]
  if (synthToken == null) return null
  return event
    .getBlock()
    .then(({ timestamp }) => {
      return {
        amount: withdraw.synthAmount,
        itemId: withdraw.itemId.toString(),
        synthToken,
        timestamp, // in seconds
        transactionHash: event.transactionHash,
        type: "withdraw",
      } as WithdrawEvent
    })
    .catch(() => null)
}

enum BridgePendingSwapTypes {
  Null,
  TokenToSynth,
  SynthToToken,
  TokenToToken,
}
async function fetchPendingSwapInfo(
  bridgeContract: Bridge,
  event: Event,
  tokens: BasicTokens,
) {
  let result = null
  try {
    // this will throw if the itemId has already fully resolved
    // afik we don't have a better way of fetching current pendingSwaps
    // DEVELOPER: for this reason we don't use multicall
    const pendingSwap = event.args as unknown as BridgeEventGenericSwap
    if (pendingSwap == null || !tokens) return null
    const itemId = pendingSwap.itemId.toString()
    const [pendingSwapInfo, eventBlock] = await Promise.all([
      bridgeContract.getPendingSwapInfo(itemId),
      event.getBlock(),
    ])
    const pendingSwapType = pendingSwapInfo.swapType
    let swapType
    if (pendingSwapType === BridgePendingSwapTypes.TokenToSynth) {
      swapType = SWAP_TYPES.TOKEN_TO_SYNTH
    } else if (pendingSwapType === BridgePendingSwapTypes.SynthToToken) {
      swapType = SWAP_TYPES.SYNTH_TO_TOKEN
    } else if (pendingSwapType === BridgePendingSwapTypes.TokenToToken) {
      swapType = SWAP_TYPES.TOKEN_TO_TOKEN
    } else {
      swapType = SWAP_TYPES.INVALID
    }
    const synthTokenFrom = tokens[pendingSwapInfo.synth]
    const tokenTo = tokens[pendingSwapInfo.tokenTo]
    if (!synthTokenFrom || !tokenTo) return null
    const settleableAtTimestamp =
      (eventBlock.timestamp + parseInt(pendingSwapInfo.secsLeft.toString())) * // add event block timestamp + secsLeft
      1000 // convert to ms
    let secondsRemaining = Math.ceil(
      (settleableAtTimestamp - Date.now()) / 1000,
    )
    secondsRemaining = Math.max(secondsRemaining, 0)
    result = {
      swapType,
      settleableAtTimestamp,
      secondsRemaining,
      synthBalance: pendingSwapInfo.synthBalance,
      itemId,
      synthTokenFrom,
      tokenTo,
      transactionHash: event.transactionHash,
      timestamp: eventBlock.timestamp,
    }
  } catch {
    // do nothing because this is probably okay
  }
  return result
}

function removeItemFromState(state: State, itemId: string): State {
  return {
    settlements: omit(state.settlements, itemId),
    withdraws: omit(state.withdraws, itemId),
    pendingSwaps: state.pendingSwaps.filter(({ itemId }) => itemId !== itemId),
  }
}

export default usePendingSwapData
