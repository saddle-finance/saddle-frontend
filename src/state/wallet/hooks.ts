import { BLOCK_TIME, ChainId, TOKENS_MAP } from "../../constants"
import { Contract, Provider } from "ethcall"
import { MulticallContract, MulticallProvider } from "../../types/ethcall"

import { BigNumber } from "@ethersproject/bignumber"
import ERC20_ABI from "../../constants/abis/erc20.json"
import { Erc20 } from "../../../types/ethers-contracts/Erc20"
import { useActiveWeb3React } from "../../hooks"
import usePoller from "../../hooks/usePoller"
import { useState } from "react"

export function usePoolTokenBalances(): { [token: string]: BigNumber } | null {
  const { account, chainId, library } = useActiveWeb3React()
  const [balances, setBalances] = useState<{ [token: string]: BigNumber }>({})

  const ethcallProvider = new Provider() as MulticallProvider

  usePoller((): void => {
    async function pollBalances(): Promise<void> {
      if (!library || !chainId || !account) return

      await ethcallProvider.init(library)
      // override the contract address when using hardhat
      if (chainId == ChainId.HARDHAT) {
        ethcallProvider.multicallAddress =
          "0xa85233C63b9Ee964Add6F2cffe00Fd84eb32338f"
      }

      const tokens = Object.values(TOKENS_MAP)
      const balanceCalls = tokens
        .map((t) => {
          return new Contract(
            t.addresses[chainId],
            ERC20_ABI,
          ) as MulticallContract<Erc20>
        })
        .map((c) => c.balanceOf(account))
      const balances = await ethcallProvider.all(balanceCalls, {})

      const ethBalance = await library.getBalance(account)
      setBalances(
        tokens.reduce(
          (acc, t, i) => ({
            ...acc,
            [t.symbol]: balances[i],
          }),
          { ETH: ethBalance },
        ),
      )
    }
    if (account) {
      void pollBalances()
    }
  }, BLOCK_TIME)

  return balances
}
