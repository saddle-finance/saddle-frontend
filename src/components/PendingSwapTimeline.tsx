import { Link, SvgIconTypeMap, styled, useTheme } from "@mui/material"
import {
  TimelineItem as MuiTimelineItem,
  Timeline,
  TimelineConnector,
  TimelineContent,
  TimelineDot,
  TimelineSeparator,
} from "@mui/lab"
import React, { ReactElement } from "react"
import ArrowDownIcon from "@mui/icons-material/ArrowDownward"
import CheckIcon from "@mui/icons-material/Check"
import ClockIcon from "@mui/icons-material/AccessTime"
import { OverridableComponent } from "@mui/material/OverridableComponent"
import { PendingSwap } from "../hooks/usePendingSwapData"
import { commify } from "../utils"
import { formatBNToString } from "../utils"
import { getFormattedShortTime } from "../utils/dateTime"
import { useTranslation } from "react-i18next"

const TimelineItem = styled(MuiTimelineItem)(() => ({
  "&:before": {
    flex: "unset",
    padding: 0,
  },
}))
const PendingSwapTimeline = ({
  pendingSwap,
}: {
  pendingSwap: PendingSwap
}): ReactElement => {
  const { t } = useTranslation()
  const {
    secondsRemaining,
    tokenTo,
    synthBalance,
    synthTokenFrom,
    timestamp,
    transactionHash,
    events,
  } = pendingSwap
  const minutesRemaining = Math.max(Math.ceil(secondsRemaining / 60), 0)
  if (!tokenTo || !synthTokenFrom) return <></>
  const formattedBalance = commify(
    formatBNToString(synthBalance, synthTokenFrom.decimals, 6),
  )
  const hasEvents = events.length > 0

  return (
    <Timeline position="right" sx={{ padding: 0 }}>
      <TimelineStep
        isActive={false}
        icon={CheckIcon}
        testId="PendingSwapTimeline:step1Of2"
      >
        {getFormattedShortTime(timestamp)} {t("stepAOfB", { a: 1, b: 2 })}
        {" - "}
        {t("confirmTheSwap")}{" "}
        <Link
          href={`https://etherscan.io/tx/${transactionHash}`}
          target="_blank"
          rel="noreferrer"
        >
          {transactionHash.slice(0, 8)}
        </Link>
      </TimelineStep>
      {minutesRemaining > 0 ? (
        <TimelineStep
          isActive={true}
          icon={ClockIcon}
          testId="PendingSwapTimeline:minutesLeft"
        >
          {t("minutesLeft", { count: minutesRemaining })}
        </TimelineStep>
      ) : (
        <TimelineStep
          isActive={!hasEvents}
          icon={CheckIcon}
          testId="PendingSwapTimeline:swappedForAmount"
        >
          {getFormattedShortTime(timestamp)}{" "}
          {t("swappedForAmount", { amount: formattedBalance })}{" "}
          {synthTokenFrom.symbol}
        </TimelineStep>
      )}
      {!hasEvents ? (
        <TimelineStep
          isActive={false}
          icon={ArrowDownIcon}
          testId="PendingSwapTimeline:step2Of2"
          withLine={false}
        >
          {t("stepAOfB", { a: 2, b: 2 })}{" "}
          {t("settleToken", { name: tokenTo.symbol })}{" "}
        </TimelineStep>
      ) : null}
      {events.map((event, i) => {
        if (event.type === "settlement") {
          return (
            <TimelineStep
              isActive={false}
              icon={CheckIcon}
              key={event.transactionHash}
              testId={`PendingSwapTimeline:event-${i + 1}`}
              withLine={false}
            >
              {getFormattedShortTime(event.timestamp)}{" "}
              {t("settledXTokenForYToken", {
                amountFrom: commify(
                  formatBNToString(
                    event.fromAmount,
                    event.fromToken.decimals,
                    6,
                  ),
                ),
                nameFrom: event.fromToken.symbol,
                amountTo: commify(
                  formatBNToString(event.toAmount, event.toToken.decimals, 6),
                ),
                nameTo: event.toToken.symbol,
              })}
            </TimelineStep>
          )
        } else if (event.type === "withdraw") {
          const formattedAmount = commify(
            formatBNToString(event.amount, event.synthToken.decimals, 6),
          )
          return (
            <TimelineStep
              isActive={false}
              icon={CheckIcon}
              key={i}
              testId={`PendingSwapTimeline:event-${i + 1}`}
              withLine={false}
            >
              {getFormattedShortTime(event.timestamp)}{" "}
              {t("withdrewNToken", {
                amount: formattedAmount,
                name: event.synthToken.symbol,
              })}
            </TimelineStep>
          )
        }
      })}
    </Timeline>
  )
}
export default PendingSwapTimeline

const TimelineStep = ({
  isActive,
  icon: Icon,
  children,
  testId,
  withLine = true,
}: React.PropsWithChildren<{
  isActive: boolean
  icon: OverridableComponent<SvgIconTypeMap>
  testId?: string
  withLine?: boolean
}>) => {
  const theme = useTheme()
  return (
    <TimelineItem data-testid={testId || null}>
      <TimelineSeparator>
        <TimelineDot
          sx={{
            backgroundColor: isActive
              ? theme.palette.primary.main
              : theme.palette.text.secondary,
            width: 12,
            height: 12,
          }}
        />
        {withLine && (
          <TimelineConnector
            sx={{ background: theme.palette.text.secondary }}
          />
        )}
      </TimelineSeparator>
      {/* TimelineContent is extended Typography so props of Typography works for this component */}
      <TimelineContent
        variant={isActive ? "subtitle2" : "body2"}
        color={isActive ? "primary" : "text.secondary"}
      >
        <Icon
          sx={{
            height: 16,
            width: 16,
            marginRight: 1,
            marginBottom: "-4px",
          }}
        />
        {children}
      </TimelineContent>
    </TimelineItem>
  )
}
