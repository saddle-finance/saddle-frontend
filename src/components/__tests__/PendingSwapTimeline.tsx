import "@testing-library/jest-dom/extend-expect"

import { SWAP_TYPES, TOKENS_MAP } from "../../constants"
import { render, screen } from "@testing-library/react"

import { BigNumber } from "@ethersproject/bignumber"
import { PendingSwap } from "../../hooks/usePendingSwapData"
import PendingSwapTimeline from "../PendingSwapTimeline"
import React from "react"

const eventTime = Math.floor(new Date(2021, 1, 1).getTime() / 1000)
const secondsRemaining = 60 * 5
const basePendingSwap: PendingSwap = {
  swapType: SWAP_TYPES.INVALID,
  settleableAtTimestamp: eventTime + secondsRemaining,
  secondsRemaining,
  synthTokenFrom: TOKENS_MAP["sETH"],
  synthBalance: BigNumber.from("0"),
  tokenTo: TOKENS_MAP["WBTC"],
  itemId: "1",
  transactionHash: "0x0",
  timestamp: eventTime,
  events: [],
}

const buildPendingSwap = (partialSwap: Partial<PendingSwap>) => {
  return {
    ...basePendingSwap,
    ...partialSwap,
  }
}

jest.mock("react-i18next", () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => null),
      },
    }
  },
}))

it("renders a basic swap correctly", () => {
  const { asFragment } = render(
    <PendingSwapTimeline pendingSwap={basePendingSwap} />,
  )
  expect(
    screen.getByTestId("PendingSwapTimeline:minutesLeft"),
  ).toHaveTextContent("minutesLeft")
  expect(asFragment()).toMatchSnapshot()
})

it("renders a swap after the waiting period correctly", () => {
  const swap = buildPendingSwap({ secondsRemaining: 0 })
  const { asFragment } = render(<PendingSwapTimeline pendingSwap={swap} />)
  expect(asFragment()).toMatchSnapshot()
})

it("renders a swap with events correctly", () => {
  const settleEvent = {
    timestamp: eventTime + 10 * 60,
    itemId: "1",
    transactionHash: "0x1",
    type: "settlement" as const,
    fromToken: TOKENS_MAP["sBTC"],
    fromAmount: BigNumber.from(10).pow(17),
    toToken: TOKENS_MAP["WBTC"],
    toAmount: BigNumber.from(10).pow(17),
  }
  const withdrawEvent = {
    timestamp: eventTime + 11 * 60,
    itemId: "1",
    transactionHash: "0x2",
    type: "withdraw" as const,
    synthToken: TOKENS_MAP["sBTC"],
    amount: BigNumber.from(10).pow(17),
  }
  const swapWithEvents = buildPendingSwap({
    events: [settleEvent, withdrawEvent],
  })
  const { asFragment } = render(
    <PendingSwapTimeline pendingSwap={swapWithEvents} />,
  )
  expect(screen.getByTestId("PendingSwapTimeline:event-1")).toHaveTextContent(
    "settled",
  )
  expect(screen.getByTestId("PendingSwapTimeline:event-2")).toHaveTextContent(
    "withdrew",
  )

  expect(asFragment()).toMatchSnapshot()
})
