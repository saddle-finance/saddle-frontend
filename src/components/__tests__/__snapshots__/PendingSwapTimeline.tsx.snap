// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a basic swap correctly 1`] = `
<DocumentFragment>
  <div
    class="timeline"
  >
    <div
      class="timelineStep inactive"
      data-testid="PendingSwapTimeline:step1Of2"
    >
      <svg>
        check.svg
      </svg>
       2/1 12:00 AM stepAOfB confirmTheSwap 
      <a
        href="https://etherscan.io/tx/0x0"
        rel="noreferrer"
        target="_blank"
      >
        0x0
      </a>
    </div>
    <div
      class="timelineStep active"
      data-testid="PendingSwapTimeline:minutesLeft"
    >
      <div
        class="timelineLine"
      />
      <svg>
        clock.svg
      </svg>
       minutesLeft
    </div>
    <div
      class="timelineStep inactive"
      data-testid="PendingSwapTimeline:step2Of2"
    >
      <div
        class="timelineLine"
      />
      <svg>
        arrowDown.svg
      </svg>
       stepAOfB settleToken 
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders a swap after the waiting period correctly 1`] = `
<DocumentFragment>
  <div
    class="timeline"
  >
    <div
      class="timelineStep inactive"
      data-testid="PendingSwapTimeline:step1Of2"
    >
      <svg>
        check.svg
      </svg>
       2/1 12:00 AM stepAOfB confirmTheSwap 
      <a
        href="https://etherscan.io/tx/0x0"
        rel="noreferrer"
        target="_blank"
      >
        0x0
      </a>
    </div>
    <div
      class="timelineStep active"
      data-testid="PendingSwapTimeline:swappedForAmount"
    >
      <div
        class="timelineLine"
      />
      <svg>
        check.svg
      </svg>
       2/1 12:00 AM swappedForAmount sETH
    </div>
    <div
      class="timelineStep inactive"
      data-testid="PendingSwapTimeline:step2Of2"
    >
      <div
        class="timelineLine"
      />
      <svg>
        arrowDown.svg
      </svg>
       stepAOfB settleToken 
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders a swap with events correctly 1`] = `
<DocumentFragment>
  <div
    class="timeline"
  >
    <div
      class="timelineStep inactive"
      data-testid="PendingSwapTimeline:step1Of2"
    >
      <svg>
        check.svg
      </svg>
       2/1 12:00 AM stepAOfB confirmTheSwap 
      <a
        href="https://etherscan.io/tx/0x0"
        rel="noreferrer"
        target="_blank"
      >
        0x0
      </a>
    </div>
    <div
      class="timelineStep active"
      data-testid="PendingSwapTimeline:minutesLeft"
    >
      <div
        class="timelineLine"
      />
      <svg>
        clock.svg
      </svg>
       minutesLeft
    </div>
    <div
      class="timelineStep inactive"
      data-testid="PendingSwapTimeline:event-1"
    >
      <div
        class="timelineLine"
      />
      <svg>
        check.svg
      </svg>
       2/1 12:10 AM settledXTokenForYToken
    </div>
    <div
      class="timelineStep inactive"
      data-testid="PendingSwapTimeline:event-2"
    >
      <div
        class="timelineLine"
      />
      <svg>
        check.svg
      </svg>
       2/1 12:11 AM withdrewNToken
    </div>
  </div>
</DocumentFragment>
`;
